### DISCUSSION TITLE
Is there Deno VM like NodeJS.vm ?

### DISCUSSION THREAD

**Discussion by gilvex** [denoland/deno#11269] [Q&A] [answered]
Hi! In NodeJS there was module "vm" which allowed you to run untrusted code in new context 
The one limit was that you couldn't run TYPESCRIPT code in NodeJS's vm

But I didn't find any of that in Deno. Is there a way to run untrusted TYPESCRIPT code in Deno with a context?

**Comment by lucacasonato** [comment, marked as answer]
The "vm" module in Node can **not** be used to run untrusted code. **It is not a security mechanism!** Let me repeat,

## The "vm" module in Node is NOT a security mechanism! Do not use it to run untrusted code! If you do this in production, shut down the server right now.

Read the first line in the Node.js docs: https://nodejs.org/api/vm.html#vm_vm_executing_javascript

---

Deno can run untrusted code in [web workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) if you constrain their permissions. They make for a more adequate sandbox. With just web workers you might be vulnerable to in process timing attacks. To prevent this, use deno subprocesses instead of workers.

**Reply by gilvex** [reply]
Is there a way to add Context in WebWorkers or Subprocess of Deno? I'm still searching about it and didn't find anything about setting custom Context, Globals to them

**Reply by lucacasonato** [reply]
No, this is not possible.

**Reply by gilvex** [reply]
Is there a way to run untrusted typescript code with a custom context in Deno?

**Reply by lucacasonato** [reply]
No - don't use contexts to run untrusted code.

If you want a custom runtime with custom builtins, you can build one in Rust by utilizing `deno_core`.

**Reply by kitsonk** [reply]
What is the use case?

If you want to adjust the context before you run unstrusted code, you can adjust whatever you want and then do an `import("./my_untrusted_code.ts")`.  If it is truly dynamic, you can use a `data` URL or a `blob` URL to dynamically import the code.

**Reply by gilvex** [reply]
> What is the use case?
> 
> If you want to adjust the context before you run unstrusted code, you can adjust whatever you want and then do an `import("./my_untrusted_code.ts")`. If it is truly dynamic, you can use a `data` URL or a `blob` URL to dynamically import the code.

The deal is that you can't do it with this way. If the project is built to binary then you can't use import
![image](https://user-images.githubusercontent.com/50551032/124428927-e9e5dd80-dd75-11eb-84af-3255a77ea308.png)
![image](https://user-images.githubusercontent.com/50551032/124429100-1e599980-dd76-11eb-8970-7861cc1b9912.png)
![image](https://user-images.githubusercontent.com/50551032/124429143-2a455b80-dd76-11eb-8bb0-f819d25f4ab3.png)

**Reply by gilvex** [reply]
Btw I'm struggling with building custom runtime in rust with deno_core Is there any information to learn about this?

**Reply by gilvex** [reply]
anyone? 😥
